TCPStore add watchKey method and new listener thread

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56217 TCPStore add watchKey method and new listener thread**

Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update utils sendBytes() to include flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already captured
- Update watchKey() to block until callback has also been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran build/bin/TCPStoreTest 100 times locally with no errors, ran [ci-all tests](https://github.com/pytorch/pytorch/pull/56219) (passed all)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

H-Huang committed:
TCPStore add watchKey method and new listener thread

[ghstack-poisoned]

H-Huang committed:
Update on "[Reland] TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Fix race condition in testWatchKeyCallback which caused flaky test failures.


[ghstack-poisoned]

H-Huang committed:
Update on "[Reland] TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update watchKey to block until callback has been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Ran TCPStoreTest 100 times locally with no errors

[ghstack-poisoned]

H-Huang committed:
Update on "[Reland] TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update socket send() to use flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already capturad
- Update watchKey to block until callback has been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran TCPStoreTest 100 times locally with no errors, running ci-all tests

[ghstack-poisoned]

H-Huang committed:
Update on "[Reland] TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update socket send() to use flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already capturad
- Update watchKey to block until callback has been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran TCPStoreTest 100 times locally with no errors, running ci-all tests

[ghstack-poisoned]

H-Huang committed:
Update on "[Reland] TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update socket send() to use flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already capturad
- Update watchKey to block until callback has been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran TCPStoreTest 100 times locally with no errors, running [ci-all tests](https://github.com/pytorch/pytorch/pull/56219)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update socket send() to use flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already capturad
- Update watchKey() to block until callback has been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran TCPStoreTest 100 times locally with no errors, running [ci-all tests](https://github.com/pytorch/pytorch/pull/56219)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
+      const std::string& key,
+      const enum WatchResponseType& type,
+      std::vector<uint8_t>& oldData,
+      std::vector<uint8_t>& newData);
```
Curious, why these functions are protected instead of private? Do we expect any other class inherit from `TCPStoreMasterDaemon`?

H-Huang replied:
They were previously set as protected, but I will change them all to private since we don't expect inheritance

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
+
+ protected:
+  void run();
+  void callbackHandler(int socket);
```
ditto

mrshenli commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -582,10 +581,15 @@ void sendBytes(
   }
 #endif
 
+#ifdef MSG_NOSIGNAL
```
shall we add comments and references to explain why setting this bit?

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,11 +56,174 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+
+void TCPStoreWorkerDaemon::addCallback(
+    std::string key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)>
```
nit: we can create an alias of this callback function with `using`.

H-Huang replied:
Thanks!

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -522,6 +727,27 @@ bool TCPStore::deleteKey(const std::string& key) {
+
+  // Register callback with TCPStoreMasterDaemon to call TCPStoreWorkerDaemon on
+  // key change
+  tcpStoreWorkerDaemon_->addCallback(regKey, callback);
+  tcputil::sendValue<QueryType>(listenSocket_, QueryType::WATCH_KEY);
+  tcputil::sendString(listenSocket_, regKey);
```
Sending the request is not protected by a lock. Does this mean it is possible to send multiple register callback requests concurrently? If that's true, it also means that the following critical section can match with a different request? E.g., say we have two registering requests, and the response of the 2nd request comes back first and unblocks the first request's CV. Does this mean when `watchKey` returns, it does not always guarantee that the registration is done on the master?

H-Huang replied:
Updated the TCPStore::watchKey method to use a lock so only one register callback request can be sent per process at a time.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,11 +56,174 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)>
+        callback) {
+  const std::lock_guard<std::mutex> lock(keyToCallbacksMutex);
+  keyToCallbacks[key] = callback;
```
do we allow registering callbacks for the same key multiple times (overriding)?

H-Huang replied:
Yes. Only one callback can be set for each key and it is overriden if set again.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,11 +56,174 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+  auto watchResponse = tcputil::recvValue<WatchResponseType>(socket);
+  if (watchResponse == WatchResponseType::KEY_CALLBACK_REGISTERED) {
+    // Notify the waiting "watchKey" operation to return
+    callbackRegistered = true;
```
this can be accessed by multiple threads? 

1. callbackHandler can set this to true
2. TCPStore::watchKey can set this to false concurrently?

H-Huang replied:
Great point, I will add a lock to TCPStore::watchkey

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
+      std::function<
+          void(c10::optional<std::string>, c10::optional<std::string>)> cb);
+  std::mutex mtx;
+  std::condition_variable cv;
+  bool callbackRegistered = false;
```
it's a bit weird to have public fields like this. IIUC, this is to allow `TCPStore` to easily manipulate `TCPStoreWorkerDaemon` thread states. Is it possible to wrap those into `TCPStoreWorkerDaemon` member functions so that we don't need to expose its raw internal states.

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update utils sendBytes() to include flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already captured
- Update watchKey() to block until callback has also been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran build/bin/TCPStoreTest 100 times locally with no errors, ran [ci-all tests](https://github.com/pytorch/pytorch/pull/56219) (passed all)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/Store.hpp:
```diff
@@ -10,6 +10,11 @@
 
+// callback function will be given arguments (optional<string> oldValue,
+// optional<string> newValue)
+using StoreCallbackFunction =
```
StoreCallbackFunction -> WatchKeyCallback?

mrshenli commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -10,19 +11,27 @@
+  const auto numThreads = 16;
+  const auto numWorkers = numThreads + 1;
+
+  auto serverTCPStore = _createServer(numWorkers);
```
Would I be correct if I assume the only change in this test function is extracting the server creation logic into a dedicated `_createServer` function? Other changes are mostly clang-format? (except push_back -> emplace_back)

H-Huang replied:
Yep, that's correct those are the only changes

mrshenli commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -136,10 +151,221 @@ void testHelper(const std::string& prefix = "") {
+  for (auto i = 0; i < numThreads; i++) {
+    serverStore->watchKey(internalKey + std::to_string(i), callback);
+    serverStore->watchKey(
+        internalKey + "counter" + std::to_string(i), callback);
```
IIUC, this means that all callbacks are installed by the server. This test mainly covers the case where clients modifies keys and server callbacks are invoked correctly. Am I correct?

H-Huang replied:
Yes all the callbacks are added to the process hosting the server. That is a good point, I will modify this test to also register callbacks on all clients and server.

mrshenli commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -136,10 +151,221 @@ void testHelper(const std::string& prefix = "") {
+TEST(TCPStoreTest, testKeyEmptyUpdate) {
+  auto store = _createServer();
+
+  std::string key = "testEmptyKeyValue";
+  c10d::test::set(*store, key, "");
+  store->get(key);
+  testKeyChangeHelper(*store, key, "", "2");
+}
+
+TEST(TCPStoreTest, testKeyUpdate) {
+  auto store = _createServer();
+
+  std::string key = "testRegularKeyValue";
+  c10d::test::set(*store, key, "1");
+  store->get(key);
+  testKeyChangeHelper(*store, key, "1", "2");
+}
+
+TEST(TCPStoreTest, testKeyCreate) {
+  auto store = _createServer();
+
+  std::string key = "testWatchKeyCreate";
+  testKeyChangeHelper(*store, key, c10::nullopt, "2");
+}
+
+TEST(TCPStoreTest, testKeyAdd) {
+  auto store = _createServer();
+
+  std::string key = "testWatchKeyAdd";
+  testKeyChangeHelper(*store, key, c10::nullopt, "2");
+}
+
+TEST(TCPStoreTest, testKeyDelete) {
+  auto store = _createServer();
+
+  std::string key = "testWatchKeyDelete";
+  c10d::test::set(*store, key, "1");
+  store->get(key);
+  testKeyChangeHelper(*store, key, "1", c10::nullopt);
+}
```
Looks like these tests are all testing server stores callbacks? Do we have tests covering the case where callbacks are installed on client stores?

H-Huang replied:
Yes, these are all just for one process (acts as both server and client to itself) to simplify things since I wanted to test the correctness of the returned callback arguments for each different operation. I will update the multiple process test to add callbacks to server along with all clients as you mentioned in the previous comment!

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
+  void run();
+  void callbackHandler(int socket);
+  // List of callbacks map each watched key
+  std::unordered_map<std::string, StoreCallbackFunction> keyToCallbacks;
+  std::mutex keyToCallbacksMutex;
+  std::condition_variable callbackRegistered;
+  bool callbackRegisteredData = false;
```
nit on naming convention: other part of the code base usually adds a `_` suffix to private variable names.

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
+ public:
+  explicit TCPStoreWorkerDaemon(int listenSocket);
+  // Adds a callback to run key change
+  void addCallback(
```
curious, why do we need `addCallback` and `watchKey` as two separate APIs? Do we expect users directly call `addCallback`? Or is this mainly for testing purpose?

H-Huang replied:
the map from key -> callback is stored on TCPStoreWorkerDaemon and `addCallback` is a helper method to add to it. `watchKey` is what registers the callback (add socket to TCPStoreMasterDaemon and callback to TCPStoreWorkerDaemon) so `watchKey` is the only API that should be called

mrshenli replied:
I see, does this mean `addCallback` should be private?

H-Huang replied:
I wanted to keep the keyToCallback_ map private in TCPWorkerDaemon but still need to update it from the user process, so I need the method to be public. I will rename it to `setCallback` to be clearer

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -522,6 +726,30 @@ bool TCPStore::deleteKey(const std::string& key) {
+    const std::string& key,
+    StoreCallbackFunction callback) {
+  // Only allow one thread to perform watchKey() at a time
+  const std::lock_guard<std::mutex> watchKeyLock(watchKeyMutex);
```
curious, what is the motivation of having two mutex/locks? (Another one is `keyToCallbacksMutex`)

H-Huang replied:
keyToCallbacksMutex is for the map on TCPStoreWorkerDaemon and watchKeyMutex is for the operation of registering on masterDaemon and workerDaemon

mrshenli replied:
Trying to understand the benefit of having two locks instead of using one for both cases. Is the main motivation to allow `TCPStoreWorkerDaemon::callbackHandler` to run concurrently with this function as much as possible?

mrshenli replied:
addressed offline.

rohan-varma commented on torch/lib/c10d/Store.hpp:
```diff
@@ -55,6 +60,20 @@ class Store : public torch::CustomClassHolder {
+  virtual void watchKey(
+      const std::string& /* unused */,
+      StoreCallbackFunction /* unused */) {
+    TORCH_INTERNAL_ASSERT(false, "watchKey only implemented for TCPStore.");
```
nit: use torch_check because torch_internal_assert is for library errors/indicative of a bug? 
Also, looks like it is also implemented for PrefixStore?

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update utils sendBytes() to include flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already captured
- Update watchKey() to block until callback has also been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran build/bin/TCPStoreTest 100 times locally with no errors, ran [ci-all tests](https://github.com/pytorch/pytorch/pull/56219) (passed all)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update utils sendBytes() to include flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already captured
- Update watchKey() to block until callback has also been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran build/bin/TCPStoreTest 100 times locally with no errors, ran [ci-all tests](https://github.com/pytorch/pytorch/pull/56219) (passed all)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -522,6 +726,28 @@ bool TCPStore::deleteKey(const std::string& key) {
+  // Block until callback has been registered successfully
+  std::unique_lock<std::mutex> callbackRegistrationLock(
+      callbackRegistrationMutex);
+  tcpStoreWorkerDaemon_->getCallbackRegisteredCondVar().wait(
+      callbackRegistrationLock,
+      [&] { return tcpStoreWorkerDaemon_->getCallbackRegisteredData(); });
+
+  // Reset payload for next callback
+  tcpStoreWorkerDaemon_->setCallbackRegisteredData(false);
```
sorry for misleading you. My original intention is to wrap these two lines into one function in `tcpStoreWorkerDaemon_`, so that we don't need to expose its internal CVs.

H-Huang replied:
Gotcha, made the change!

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,11 +56,173 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+  auto watchResponse = tcputil::recvValue<WatchResponseType>(socket);
+  if (watchResponse == WatchResponseType::KEY_CALLBACK_REGISTERED) {
+    // Notify the waiting "watchKey" operation to return
+    setCallbackRegisteredData(true);
```
this can run concurrently with another invocation in `TCPStore::watchKey`. Will this be a problem? Say it is set to true here, and then immediately flipped to false by `TCPStore::watchKey`. And then the next line `getCallbackRegisteredCondVar().notify_one();` kicks in?

mrshenli replied:
addressed offline. This won't occur, as TCPStore::watchKey will block until this is done before processing the confirmation from the master.

mrshenli reviewed (commented):
One thing that can be helpful is to have a comment paragraph to argue the correctness of how we handle watch callbacks concurrently.

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update utils sendBytes() to include flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already captured
- Update watchKey() to block until callback has also been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran build/bin/TCPStoreTest 100 times locally with no errors, ran [ci-all tests](https://github.com/pytorch/pytorch/pull/56219) (passed all)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/PrefixStore.cpp:
```diff
@@ -46,6 +46,10 @@ bool PrefixStore::deleteKey(const std::string& key) {
 }
 
+void PrefixStore::watchKey(const std::string& key, WatchKeyCallback callback) {
+  return store_->watchKey(joinKey(key), callback);
```
`std::move(callback)`?

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +70,58 @@ class TCPStoreDaemon {
+    callbackRegisteredData_ = false;
+  }
+  void setCallbackRegistered() {
+    callbackRegisteredData_ = true;
```
do we need `callbackRegistrationMutex_` to protect this set here? I guess it is not required today, because `TCPStore::watchKey` will block until `callbackRegisteredData_` is set to `True`. So the second watchKey request won't be sent until the `nofity_one()` below is called?

H-Huang replied:
Yep, that is correct. Not required since only one watchKey request can be sent at a time.

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -105,15 +175,17 @@ class TCPStore : public Store {
       const std::vector<std::string>& keys,
       const std::chrono::milliseconds& timeout);
 
+  std::mutex watchKeyMutex;
```
`watchKeyMutex` -> `watchKeyMutex_`

mrshenli commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -136,10 +151,222 @@ void testHelper(const std::string& prefix = "") {
+  for (auto i = 0; i < numThreads; i++) {
+    serverStore->watchKey(internalKey + std::to_string(i), callback);
+    serverStore->watchKey(internalKeyCount + std::to_string(i), callback);
+    clientStores[i]->watchKey(internalKey + std::to_string(i), callback);
+    clientStores[i]->watchKey(internalKeyCount + std::to_string(i), callback);
```
these above two lines are newly added to test registering watch key from client side, right?

H-Huang replied:
Yep!

mrshenli reviewed (approved):
LGTM!

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


Reland of https://github.com/pytorch/pytorch/pull/54264

Changes:
- Update utils sendBytes() to include flag MSG_NOSIGNAL to prevent SIGPIPE because error in return is already captured
- Update watchKey() to block until callback has also been registered on master.
- Fix race condition in testWatchKeyCallback which caused flaky test failures.

Test:
Ran build/bin/TCPStoreTest 100 times locally with no errors, ran [ci-all tests](https://github.com/pytorch/pytorch/pull/56219) (passed all)

Differential Revision: [D27824802](https://our.internmc.facebook.com/intern/diff/D27824802)

[ghstack-poisoned]
